text/0000-iterator-try_all-try_any.md

```diff
@@ -0,0 +1,144 @@
+
+- a `for` loop is required to iterate over the elements and stop early on errors or invalid items
+- collecting the elements in the iterator to a collected `Result`, then calling `all` on the inner collection
+
+Both solutions are not very elegant. The `try_all` method would work as `.try_all(check)`, returning `Result<bool, Error>` as well. It should behave similar to `all`, except working with `Try<bool>` output, just as `try_fold` behaves in comparison to `fold`.
+
+Everything applies similarly to `try_any`.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+These additions of `try_all` and `try_any` to the core library of Rust allow the use of fallible validators in iterators.
+
+When every element in an iterator needs to be validated, it could be done like this:
+
+```rust
+fn file_is_empty(path: &str) -> std::io::Result<bool> {
+  Ok(std::fs::read_to_string(path)?.is_empty())
+}
+
+fn main() {
+    let files = ["fileA.txt", "fileB.txt"];
+    let all_exist_but_empty = files.iter().try_all(file_is_empty).expect("file wasn't found");
+    let any_exists_but_empty = files.iter().try_any(file_is_empty).expect("file wasn't found");
+}
+```
+
+It works similarly with `Option`s, e.g. the mentioned key-value-store:
+
+```rust
+fn is_valid(object: &MyObject) -> bool {
+    *object != MyObject::default()
+}
+
+fn main() {
+    let mut kv = HashMap::new();
+    kv.insert("id-1", MyObject::default());
+
+    let object_list = ["id-1", "id-5"];
+    let all_valid = object_list.iter().try_all(|id| kv.get(id).map(is_valid)).expect("id wasn't found");
+}
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The addition `try_all` behaves to `all` like `try_fold` behaves to `fold`. The similar applies to `try_any`. They both should react as their simple counter-part: short-circuit on elements that already determine the overall result. However two things are different:
+
+- They return the same type as the given function/closure does, so `Option<bool>` or `Result<bool, E>`.
+- They also short-circuit on try-fails.
+
+## Exemplary implementation
+
+Inside the Iterator trait:
+
+```rust
+fn try_all<F, R>(&mut self, mut f: F) -> R
+where
+    Self: Sized,
+    F: FnMut(Self::Item) -> R,
+    R: Try<Output = bool>,
+{
+    for item in self {
+        if !f(item)? {
+            return try { false };
+        }
+    }
+    try { true }
+}
+
+fn try_any<F, R>(&mut self, mut f: F) -> R
+where
+    Self: Sized,
+    F: FnMut(Self::Item) -> R,
+    R: Try<Output = bool>,
+{
+    for item in self {
+        if f(item)? {
+            return try { true };
+        }
+    }
+    try { false }
+}
+```
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+Adding new methods to the trait makes it even bigger, but I would argue that is a low cost (especially in this small case) for a nice benefit.
+
+# Rationale and alternatives
```

Comment:
## Can there be a general adapter for these?

I did recently r+ `try_collect` for nightly (https://github.com/rust-lang/rust/pull/94041), but now I'm wondering if instead of having all of these -- especially for non-adapters -- we should expose the [shunting mechanism](https://github.com/rust-lang/rust/blob/master/library/core/src/iter/adapters/mod.rs#L130-L200) that they use somehow instead.

Notably, the `.try_collect()` could plausibly be exposed as `.try_mut(|i| i.collect())`, so I wonder if there'd be a way that `try_all` and `try_find` and friends could also be exposed via a similar mechanism.  (And thus remove `try_collect` and `try_find` and such from the trait altogether.  `try_fold` is a slightly different case as it's there to be the internal iteration primitive, and is only secondarily there as something to be used directly.)

Reply:
I think this is unlikely (without major language changes).

`try_find`/`try_all`/etc have to deal with a closure, which makes them a lot more complicated than `try_collect`. If you wanted to implemented `try_find` in the shape of `.try(|i| i.find(...))` then you'd have to return a `bool` from the closure, but you have a closure that returns an `impl Try<Output = bool>`. So how do you handle a failure? There are two hurdles:
- The closure needs to communicate to the iterator that a failure has happened and iteration can stop. This probably requires having a struct that wraps the given closure and implements the `Fn*` traits. This struct can then 
- After reporting the failure, the closure still needs to return _something_ (a `bool` in this case). It doesn't even matter much what it returns, as the iteration stop before the next element anyway, but it needs to return something. A `Default` bound might solve it for most methods, but what about e.g. `try_max_by_key`?

We can almost get there, but not quite all the way it seems. Implementing non-`try` methods in terms of their `try` counterparts is, of course, a lot simpler.